runtime/src/leader_bank_notifier.rs

```diff
@@ -0,0 +1,161 @@
+    /// Current state (slot, bank, and status) of the system
+    state: Mutex<SlotAndBankWithStatus>,
+    /// CondVar to notify status changes and waiting
+    condvar: Condvar,
+}
+
+/// Leader status state machine for the validator.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+enum Status {
+    /// The leader bank is not currently available. Either not initialized, or PoH-completed bank.
+    #[default]
+    StandBy,
+    /// PoH-initiated bank is available.
+    InProgress,
+}
+
+#[derive(Debug, Default)]
+struct SlotAndBankWithStatus {
+    status: Status,
+    slot: Slot,
+    bank: Weak<Bank>,
+}
+
+impl LeaderBankNotifier {
+    /// Set the status to `InProgress` and notify any waiting threads
+    /// if the status was not already `InProgress`.
+    pub fn set_in_progress(&self, bank: &Arc<Bank>) {
+        let mut state = self.state.lock().unwrap();
+        if matches!(state.status, Status::InProgress) {
+            return;
+        }
+
+        *state = SlotAndBankWithStatus {
+            status: Status::InProgress,
+            slot: bank.slot(),
+            bank: Arc::downgrade(bank),
+        };
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// Set the status to `StandBy` and notify any waiting threads if
+    ///     1. the status was not already `StandBy` and
+    ///     2. the slot is higher than the current slot (sanity check).
+    pub fn set_completed(&self, slot: Slot) {
+        let mut state = self.state.lock().unwrap();
+        if matches!(state.status, Status::StandBy) || slot < state.slot {
+            return;
+        }
+
+        state.status = Status::StandBy;
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// If the status is `InProgress`, immediately return a weak reference to the bank.
+    /// Otherwise, wait up to the `timeout` for the status to become `InProgress`.
+    /// Returns `None` if the timeout is reached.
+    pub fn wait_for_in_progress(&self, timeout: Duration) -> Option<Weak<Bank>> {
+        let state = self.state.lock().unwrap();
+        let (state, wait_timeout_result) = self
+            .condvar
+            .wait_timeout_while(state, timeout, |state| {
+                matches!(state.status, Status::StandBy)
+            })
+            .unwrap();
+
+        (!wait_timeout_result.timed_out()).then(|| state.bank.clone())
+    }
+
+    /// Wait for next notification for a completed leader slot.
+    /// Returns `None` if the timeout is reached
+    pub fn wait_for_next_completed(&self, mut remaining_timeout: Duration) -> Option<Slot> {
+        loop {
+            let start = Instant::now();
+            let state = self.state.lock().unwrap();
+            let (state, result) = self.condvar.wait_timeout(state, remaining_timeout).unwrap();
+            if result.timed_out() {
+                return None;
+            }
+
+            if matches!(state.status, Status::StandBy) {
+                return Some(state.slot);
```

Comment:
i think `Some(0)` or any stale `Some(N)` could wrongly be returned from here due to spurious wakeup of condvars...

in short, it's not guranteed that only once wakeup is triggered for state transitions. this thread could be waken up multiple time while state.status == Status::StandBy still.

so, how about putting `let slot = ...;` outside this `loop`  for bookkeeping to guarntee progression is actually cranked?

Reply:
We, unfortunately, cannot check for slot changes because it may initially be called while already `InProgress` (same slot as it will be when we call `set_completed`)

I've taken an alternative approach where I explicitly check for state change from `InProgress` -> `StandBy`, waiting for `InProgress` if necessary.

Reply:
^ commit for above alternative approach with explicit `InProgress` -> `StandBy` transition: 58984d84223e08abc5768fb05794c9c9319ccbcb